bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -299,23 +299,36 @@ void ReplicaImp::tryToSendPrePrepareMsg(bool batchingLogic) {
   PrePrepareMsg *pp =
       new PrePrepareMsg(config_.replicaId, curView, (primaryLastUsedSeqNum + 1), firstPath, primaryCombinedReqSize);
 
-  ClientRequestMsg *nextRequest = requestsQueueOfPrimary.front();
-  while (nextRequest != nullptr && nextRequest->size() <= pp->remainingSizeForRequests()) {
-    MDC_CID_PUT(GL, nextRequest->getCid());
-    if (clientsManager->noPendingAndRequestCanBecomePending(nextRequest->clientProxyId(),
-                                                            nextRequest->requestSeqNum())) {
-      pp->addRequest(nextRequest->body(), nextRequest->size());
-      clientsManager->addPendingRequest(nextRequest->clientProxyId(), nextRequest->requestSeqNum());
+  ClientRequestMsg *nextRequest = (!requestsQueueOfPrimary.empty() ? requestsQueueOfPrimary.front() : nullptr);
+
+  while (nextRequest != nullptr) {
+    if (nextRequest->size() <= pp->remainingSizeForRequests()) {
+      MDC_CID_PUT(GL, nextRequest->getCid());
+      if (clientsManager->noPendingAndRequestCanBecomePending(nextRequest->clientProxyId(),
+                                                              nextRequest->requestSeqNum())) {
+        pp->addRequest(nextRequest->body(), nextRequest->size());
+        clientsManager->addPendingRequest(nextRequest->clientProxyId(), nextRequest->requestSeqNum());
+      }
+      primaryCombinedReqSize -= nextRequest->size();
+    } else if (pp->numberOfRequests() <= 0) {
```

Comment:
- In the current implementation probably won’t be negative value of requests.
The number of requests type is an int, I thought that it will be safer in case of unpredictable changes in the future, what do you think?

- The bug triggered once a big request was the first one to pop, in this case the request would have stuck in the queue, number of requests would have been zero and the assert in ‘finishAddingRequest()’ would have crash our replica. But I think you are right it will be helpful to log the next big requests as well. Thanks.

Reply:
* For now let's have a separate error log if `pp->numberOfRequests < 0`, as it's an invariant breach
* I suggest to log an error every time a big request fails to be added to the PP message, not just for the first one.